<launch>

    <arg name="fps_all" default="8"/>

  <!-- Front Camera Aruco Detect -->
    <arg name="front_cam" default="/usb_cam_front"/>
    <arg name="front_image" default="/teleop_cam"/>
    <arg name="front_transport" default="raw"/>
  <!-- Left Camera Aruco Detect -->
    <arg name="left_cam" default="/usb_cam_left"/>
    <arg name="left_image" default="/teleop_cam"/>
    <arg name="left_transport" default="raw"/>
  <!-- Right Camera Aruco Detect -->
    <arg name="right_cam" default="/usb_cam_right"/>
    <arg name="right_image" default="/teleop_cam"/>
    <arg name="right_transport" default="raw"/>
  <!-- Back Camera Aruco Detect -->
    <arg name="back_cam" default="/usb_cam_back"/>
    <arg name="back_image" default="/teleop_cam"/>
    <arg name="back_transport" default="raw"/>

  <node name="usb_cam_front" pkg="usb_cam" type="usb_cam_node" output="screen">
    <param name="video_device" value="/dev/video-front" />
    <param name="camera_frame_id" value="front_camera_link" />
    <param name="image_width" value="1280" />
    <param name="image_height" value="720" />
    <param name="framerate" value="$(arg fps_all)" />
    <param name="pixel_format" value="mjpeg" />
    <param name="camera_info_url" value="package://rr_auto_dock/misc/ost.yaml"/>
  </node>

  <node name="usb_cam_left" pkg="usb_cam" type="usb_cam_node" output="screen">
    <param name="video_device" value="/dev/video-left" />
    <param name="camera_frame_id" value="left_camera_link" />
    <param name="image_width" value="1280" />
    <param name="image_height" value="720" />
    <param name="framerate" value="$(arg fps_all)" />
    <param name="pixel_format" value="mjpeg" />
    <param name="camera_info_url" value="package://rr_auto_dock/misc/ost.yaml"/>
  </node>

  <node name="usb_cam_right" pkg="usb_cam" type="usb_cam_node" output="screen">
    <param name="video_device" value="/dev/video-right" />
    <param name="camera_frame_id" value="right_camera_link" />
    <param name="image_width" value="1280" />
    <param name="image_height" value="720" />
    <param name="framerate" value="$(arg fps_all)" />
    <param name="pixel_format" value="mjpeg" />
    <param name="camera_info_url" value="package://rr_auto_dock/misc/ost.yaml"/>
  </node>

  <node name="usb_cam_back" pkg="usb_cam" type="usb_cam_node" output="screen">
    <param name="video_device" value="/dev/video-back" />
    <param name="camera_frame_id" value="back_camera_link" />
    <param name="image_width" value="1280" />
    <param name="image_height" value="720" />
    <param name="framerate" value="$(arg fps_all)" />
    <param name="pixel_format" value="mjpeg" />
    <param name="camera_info_url" value="package://rr_auto_dock/misc/ost.yaml"/>
  </node>

  <node name="aruco_detect_front" pkg="rr_auto_dock" type="aruco_detect" output="screen" respawn="false">
    <param name="publish_images" value="true" />
    <param name="fiducial_len" value="0.1"/>
    <param name="dictionary" value="1"/>
    <param name="do_pose_estimation" value="true"/>
    <param name="doCornerRefinement" value="false"/>

    <!--Parameters different than opencv defaults-->
    <param name="adaptiveThreshWinSizeMax" value="23"/>
    <param name="adaptiveThreshWinSizeStep" value="10"/>
    <param name="minMarkerPerimeterRate" value="0.03"/>
    <param name="polygonalApproxAccuracyRate" value="0.1"/>
    <param name="doCornerRefinement" value="true"/>
    <param name="cornerRefinementMinAccuracy" value="0.1"/>
    <param name="maxErroneousBitsInBorderRate" value="0.35"/>
    
    <param name="image_transport" value="$(arg front_transport)"/>
    <remap from="/camera" to="$(arg front_cam)/$(arg front_image)"/>
    <remap from="/camera_info" to="$(arg front_cam)/camera_info"/>
    <remap from="/fiducial_images" to="$(arg front_cam)/fiducial_images"/>
    <remap from="/fiducial_transforms" to="$(arg front_cam)/fiducial_transforms"/>
    <remap from="/fiducial_vertices" to="$(arg front_cam)/fiducial_vertices"/>
  </node>

  <node name="aruco_detect_left" pkg="rr_auto_dock" type="aruco_detect" output="screen" respawn="false">
    <param name="publish_images" value="true" />
    <param name="fiducial_len" value="0.1"/>
    <param name="dictionary" value="8"/>
    <param name="do_pose_estimation" value="true"/>
    <param name="doCornerRefinement" value="false"/>

    <!--Parameters different than opencv defaults-->
    <param name="adaptiveThreshWinSizeMax" value="23"/>
    <param name="adaptiveThreshWinSizeStep" value="10"/>
    <param name="minMarkerPerimeterRate" value="0.03"/>
    <param name="polygonalApproxAccuracyRate" value="0.1"/>
    <param name="doCornerRefinement" value="true"/>
    <param name="cornerRefinementMinAccuracy" value="0.1"/>
    <param name="maxErroneousBitsInBorderRate" value="0.35"/>
    
    <param name="image_transport" value="$(arg left_transport)"/>
    <remap from="/camera" to="$(arg left_cam)/$(arg left_image)"/>
    <remap from="/camera_info" to="$(arg left_cam)/camera_info"/>
    <remap from="/fiducial_images" to="$(arg left_cam)/fiducial_images"/>
    <remap from="/fiducial_transforms" to="$(arg left_cam)/fiducial_transforms"/>
    <remap from="/fiducial_vertices" to="$(arg left_cam)/fiducial_vertices"/>
  </node>

  <node name="aruco_detect_right" pkg="rr_auto_dock" type="aruco_detect" output="screen" respawn="false">
    <param name="publish_images" value="true" />
    <param name="fiducial_len" value="0.1"/>
    <param name="dictionary" value="8"/>
    <param name="do_pose_estimation" value="true"/>
    <param name="doCornerRefinement" value="false"/>

    <!--Parameters different than opencv defaults-->
    <param name="adaptiveThreshWinSizeMax" value="23"/>
    <param name="adaptiveThreshWinSizeStep" value="10"/>
    <param name="minMarkerPerimeterRate" value="0.03"/>
    <param name="polygonalApproxAccuracyRate" value="0.1"/>
    <param name="doCornerRefinement" value="true"/>
    <param name="cornerRefinementMinAccuracy" value="0.1"/>
    <param name="maxErroneousBitsInBorderRate" value="0.35"/>
    
    <param name="image_transport" value="$(arg right_transport)"/>
    <remap from="/camera" to="$(arg right_cam)/$(arg right_image)"/>
    <remap from="/camera_info" to="$(arg right_cam)/camera_info"/>
    <remap from="/fiducial_images" to="$(arg right_cam)/fiducial_images"/>
    <remap from="/fiducial_transforms" to="$(arg right_cam)/fiducial_transforms"/>
    <remap from="/fiducial_vertices" to="$(arg right_cam)/fiducial_vertices"/>
  </node>

  <node name="aruco_detect_back" pkg="rr_auto_dock" type="aruco_detect" output="screen" respawn="false">
    <param name="publish_images" value="true" />
    <param name="fiducial_len" value="0.1"/>
    <param name="dictionary" value="8"/>
    <param name="do_pose_estimation" value="true"/>
    <param name="doCornerRefinement" value="false"/>

    <!--Parameters different than opencv defaults-->
    <param name="adaptiveThreshWinSizeMax" value="23"/>
    <param name="adaptiveThreshWinSizeStep" value="10"/>
    <param name="minMarkerPerimeterRate" value="0.03"/>
    <param name="polygonalApproxAccuracyRate" value="0.1"/>
    <param name="doCornerRefinement" value="true"/>
    <param name="cornerRefinementMinAccuracy" value="0.1"/>
    <param name="maxErroneousBitsInBorderRate" value="0.35"/>
    
    <param name="image_transport" value="$(arg back_transport)"/>
    <remap from="/camera" to="$(arg back_cam)/$(arg back_image)"/>
    <remap from="/camera_info" to="$(arg back_cam)/camera_info"/>
    <remap from="/fiducial_images" to="$(arg back_cam)/fiducial_images"/>
    <remap from="/fiducial_transforms" to="$(arg back_cam)/fiducial_transforms"/>
    <remap from="/fiducial_vertices" to="$(arg back_cam)/fiducial_vertices"/>
  </node>
</launch>
